crypto/dilithium/p_pqdsa_test.cc

```diff
@@ -373,65 +371,61 @@ TEST_P(PQDSAParameterTest, KAT) {
 
     size_t pk_len = GetParam().public_key_len;
     size_t sk_len = GetParam().private_key_len;
     size_t sig_len = GetParam().signature_len;
+    std::string name = GetParam().name;
 
-    // The KAT files generated by the dilithium team use the optional APIs that
-    // create a signature for a message m and append the message to the end of
-    // the signature. We only want to bring the APIs that create and verify just
-    // the signature, therefore each signature is a constant
-    // DILITHIUM3_SIGNATURE_BYTES and we truncate the KAT's signed message down,
-    // "sm" down to a constant DILITHIUM3_SIGNATURE_BYTES.
-
+    std::vector<uint8_t> pub(pk_len);
+    std::vector<uint8_t> priv(sk_len);
     std::vector<uint8_t> signature(sig_len);
-    sm.resize(sig_len);
 
-    // Convert string read from KAT to int
     size_t mlen_int = std::stoi(mlen);
+    sm.resize(sig_len);
 
-    // Here we fix the DRBG (AES-CTR) so that we are able to seed it with the
-    // seed from the KAT (testing only)
-    pq_custom_randombytes_use_deterministic_for_testing();
-    pq_custom_randombytes_init_for_testing(seed.data());
-
-    // Generate our dilithium public and private key pair
-    bssl::UniquePtr<EVP_PKEY_CTX> pctx(EVP_PKEY_CTX_new_id(EVP_PKEY_PQDSA, nullptr));
-    ASSERT_TRUE(pctx);
-    ASSERT_TRUE(EVP_PKEY_CTX_pqdsa_set_params(pctx.get(),GetParam().nid));
-    ASSERT_TRUE(EVP_PKEY_keygen_init(pctx.get()));
-    EVP_PKEY *raw = nullptr;
-    ASSERT_TRUE(EVP_PKEY_keygen(pctx.get(), &raw));
-    bssl::UniquePtr<EVP_PKEY> pkey(raw);
-
-    // Compare the expected public/secret key from KATs with generated values
-    CMP_VEC_AND_PKEY_PUBLIC(pk, pkey, pk_len);
-    CMP_VEC_AND_PKEY_SECRET(sk, pkey, sk_len);
-
-    // Generate a signature for the message
-    // We use EVP_DigestSign because dilithium supports the use of
-    // non-hash-then-sign (just like ed25519) so we first init EVP_DigestSign
-    // WITHOUT a hash function.
-    bssl::ScopedEVP_MD_CTX ctx;
-    ASSERT_TRUE(EVP_DigestSignInit(ctx.get(), nullptr, nullptr, nullptr, pkey.get()));
-    ASSERT_TRUE(EVP_DigestSign(ctx.get(), signature.data(), &sig_len, msg.data(), mlen_int));
-    EXPECT_EQ(Bytes(sm), Bytes(signature.data(), sig_len));
-    ctx.Reset();
-
-    // Verify the signature
-    ASSERT_TRUE(EVP_DigestVerifyInit(ctx.get(), nullptr, nullptr, nullptr, pkey.get()));
-    ASSERT_TRUE(EVP_DigestVerify(ctx.get(), signature.data(), sig_len, msg.data(), mlen_int));
+    // Generate key pair from seed xi and assert that public and private keys
+    // are equal to expected values from KAT
+    if (name == "MLDSA65") {
+      ASSERT_TRUE(ml_dsa_65_keypair_internal(pub.data(),priv.data(),xi.data()));
+    }
+    EXPECT_EQ(Bytes(pub), Bytes(pk));
+    EXPECT_EQ(Bytes(priv), Bytes(sk));
+
+    // Prepare m_prime = (0 || ctxlen || ctx) as in FIPS 204: Algorithm 2 line 10
+    uint8_t m_prime[257];
+    size_t m_prime_len = ctxstr.size() + 2;
```

Comment:
Summarizing an offline discussion:
`ctxstr` is an optional context string that is combined with the message to be signed. It's well defined in FIPS 204 as a byte string of 255 or fewer bytes. In fact, the signing function Algorithm 2: `ML-DSA.Sign` is responsible for ensuring that the string does not exceed this length, otherwise we return fail (see line 1 - 3 of Algorithm 2).

Line 10 of algorithm 2 describes `m_prime`, the structure we are creating; `M' = Bytestobits(0 || ctx_string_len || ctx ) || M`. You can see [from upstream](https://github.com/pq-crystals/dilithium/commit/444cdcc84eb36b66fe27b3a2529ee48f6d8150c2#diff-75ebb091b78aba35449cc4e4e3731978a8888d21ccfd47c15597518c4c556302R309) how we the external signing function pre-processes the input context string:

```
 uint8_t pre[257];
  /* Prepare pre = (0, ctxlen, ctx) */
  pre[0] = 0;
  pre[1] = ctxlen;
  memcpy(&pre[2], ctx, ctxlen);
  ```

We have to emulate this within `p_pqdsa_test.cc` as we call the internal function that expects this pre-processing to be performed.